Dependency rules (#205)
ansible-collections/overview: collection_requirements.rst

----------

* The ``tags`` field MUST be set.

> * Collection dependencies must satisfy a set of rules. See the section on `Collection Dependencies <collection_dependencies_>`_ for this.

Reviewer: Collection dependencies must meet a set of rules. See the section on `Collection Dependencies <collection_dependencies_>`_ for this.

Reviewer: I think it is easier to read the word "meet" than "satisfy" reason is that the word "meet" is a common word.

Author: If you add this as a suggestion I can merge it with attribution to you.

Reviewer: Ok

Author: Thanks!

> * Collection dependencies must satisfy a set of rules. See the section on `Collection Dependencies <collection_dependencies_>`_ for this.

Reviewer: Collection dependencies must meet a set of rules. See the section on Collection Dependencies <collection_dependencies_>_ for this.

  create new ``_info`` or ``_facts`` modules instead (for more information, refer to the `Developing modules guidelines <https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_general.html#creating-an-info-or-a-facts-module>`_)
* ``check_mode`` is supported in all ``*_info`` and ``*_facts`` modules (for more information, refer to the `Development conventions <https://docs.ansible.com/ansible/devel/dev_guide/developing_modules_best_practices.html#following-ansible-conventions>`_)

.. _collection_dependencies:

Collection Dependencies
=======================

> **Notation:** if foo.bar has a dependency on baz.bam, we say that baz.bam is the collection *dependend on*, and foo.bar is the *dependent collection*.

Author: Yep, "dependend" doesn't exist. I've rewritten this sentence and made another small change. Can you check whether this is better now?

> **Notation:** if foo.bar has a dependency on baz.bam, we say that baz.bam is the collection *dependend on*, and foo.bar is the *dependent collection*.

Reviewer 2: I would suggest using:
1. `dependency` instead of `depended on` as there are a lot of `dependency|ies` are used across the paragraph and `depended on` only once OR
2. change `dependency|ies` to `collections depended on` everywhere (though the option 1 feels a bit better to me.

Generally the combination of `depended on` and `dependent collection` confuses me a lot:) Makes me focusing each time to figure out which means what:) Can be fair for other readers, i guess

> **Notation:** if foo.bar has a dependency on baz.bam, we say that baz.bam is the collection *dependend on*, and foo.bar is the *dependent collection*.

Author: The problem is that `dependency` is the relation between two collections, while `the collection dependend on` is one of the two collections involved. I skimmed over all uses of `dependenc*`, and I think its used correctly everywhere to indicate the actual dependency (as the relation), and not the dependend on collection.

> **Notation:** if foo.bar has a dependency on baz.bam, we say that baz.bam is the collection *dependend on*, and foo.bar is the *dependent collection*.

Reviewer 2: OK but it confuses me anyway, though maybe it's my personal feature

* Collection dependencies must have a lower bound on the version which is at least 1.0.0.

  * This means that all collection dependencies have to specify lower bounds on the versions, and these lower bounds should be stable releases, and not versions of the form 0.x.y.

>   * When creating new collections where collection dependencies are also under development, you need to watch out since Galaxy checks whether dependencies exist in the required versions:

Reviewer 3: Could you please expand what you mean by `under development`? Would that be `<= 1.0.0`?

Author: Yes. This is basically about two dependent collections simultaneously reaching the 1.0.0 goal while producing versions that satisfy our rules.

Reviewer 3: Wouldn't it make sense to bar not only the collections under dev but also the ones that _depend on_ collections under dev? I mean, if the purpose of the rule is reliability/stability, I suppose that would go a long way.

Author: We only accept collections that released version 1.0.0 for inclusion anyway. Also note that this part of this PR has been in the rules for years, it only moved to a new place in this document.

Reviewer: I'm concerned with the use of the word "under development", what if that could be changed to "under developed"

Let's assume this - * When creating new collections where collection dependencies are also under developed,
What do you think?

Author: I'm not a native English speaker, but "under developed" seems to mean something completely different to me. ("in development" would be an alternative - https://english.stackexchange.com/questions/238497/which-expression-is-correct-in-development-or-under-development)

Reviewer: Oooh, interesting I just learned that now. Thanks

    #. Assume that ``foo.bar`` depends on ``foo.baz``.
    #. First release ``foo.baz`` as 1.0.0.
    #. Then modify ``foo.bar``'s ``galaxy.yml`` to specify ``'>=1.0.0'`` for ``foo.baz``.
    #. Finally release ``foo.bar`` as 1.0.0.

* The dependencies between collections included in Ansible must be valid. If a dependency is violated, the involved collections must be pinned so that all dependencies are valid again. This means that the version numbers from the previous release are kept or only partially incremented so that the resulting set of versions has no invalid dependencies.

> * If a collection has a too strict dependency for a longer time, and forces another collection to be hold back, that collection will be removed from the next major Ansible release. What "longer time" means depends on when the next Ansible major release happens. If a dependent collection prevents a new major version of a collection it depends on to be included in the next major Ansible release, the dependent collection will be removed from that major release to avoid blocking the dependent collection.

Reviewer 4: to avoid blocking the collection being depended upon

Author: ```suggestion
* If a collection has a too strict dependency for a longer time, and forces another collection to be hold back, that collection will be removed from the next major Ansible release. What "longer time" means depends on when the next Ansible major release happens. If a dependent collection prevents a new major version of a collection it depends on to be included in the next major Ansible release, the dependent collection will be removed from that major release to avoid blocking the collection being dependend upon.
```

> * If a collection has a too strict dependency for a longer time, and forces another collection to be hold back, that collection will be removed from the next major Ansible release. What "longer time" means depends on when the next Ansible major release happens. If a dependent collection prevents a new major version of a collection it depends on to be included in the next major Ansible release, the dependent collection will be removed from that major release to avoid blocking the collection being dependend upon.

Reviewer 5: ```suggestion
* If a collection has a too strict dependency for a longer time, and forces another collection depended on to be held back, that collection will be removed from the next major Ansible release. What "longer time" means depends on when the next Ansible major release happens. If a dependent collection prevents a new major version of a collection it depends on to be included in the next major Ansible release, the dependent collection will be removed from that major release to avoid blocking the collection being depended on.
```

* We strongly suggest that collections also test against the ``main`` branches of their dependencies to ensure that incompatibilities with future releases of these are detected as early as possible and can be resolved in time to avoid such problems. Collections depending on other collections must understand that they bear the risk of being removed when they do not ensure compatibility with the latest releases of their dependencies.

* Collections included in Ansible must not depend on other collections except if they satisfy one of the following cases:

>   #. They have a loose dependency on one (or more) major versions of other collections included in Ansible. For example ``ansible.netcommon: >=1.0.0``, or ``ansible.netcommon: >=2.0.0, <3.0.0``. In case the collection depended on releases a new major version outside of this version range that will be included in the next major Ansible release, the dependent collection will be removed from the next major Ansible release. The cut-off date for this is feature freeze.

Reviewer 2: ```suggestion
  #. They have a loose dependency on one (or more) major versions of other collections included in Ansible. For example, ``ansible.netcommon: >=1.0.0``, or ``ansible.netcommon: >=2.0.0, <3.0.0``. In case the collection depended on releases a new major version outside of this version range that will be included in the next major Ansible release, the dependent collection will be removed from the next major Ansible release. The cut-off date for this is feature freeze.
```

Reviewer 2: also what does `a loose dependency` mean? I googled a bit but failed to find something meaningful to me. Should we elaborate on it a bit more if the word `loose` is important here? If not, i think we should re-phrase it somehow.

>   #. They are explicitly being allowed to do by the Steering Committee.

Author: Thanks!

Examples
--------

> #. community.foo 1.2.0 has a dependency on community.bar >= 1.0.0, < 1.3.0.

Reviewer 2: ```suggestion
#. ``community.foo 1.2.0`` has a dependency on ``community.bar >= 1.0.0, < 1.3.0``.
```

   * Now ``community.bar`` creates a new release ``1.3.0``. When ``community.foo`` does not create a new release with a relaxed dependency, we have to include ``community.bar 1.2.x`` in the next Ansible release despite ``1.3.0`` being available.
   * If ``community.foo`` does not relax its dependency on ``community.bar`` for some time, ``community.foo`` will be removed from the next Ansible major release.

>    * Unfortunately ``community.bar`` has to stay at ``1.2.x`` until either ``community.bar`` is removed (in the next major release), or loosens its requirements so that newer ``community.foo 1.y.z`` releases can be included.

Reviewer 5: ```suggestion
   * Unfortunately ``community.bar`` has to stay at ``1.2.x`` until either ``community.foo`` is removed (in the next major release), or loosens its requirements so that newer ``community.bar 1.3.z`` releases can be included.
```

Reviewer 5: not sure if I understand this correctly, but that's what I'm basing this change on above ^^

>    * Now community.bar creates a new release 1.3.0. When community.foo does not create a new release with a relaxed dependency, we have to include community.bar 1.2.x in the next Ansible release despite 1.3.0 being available.
>    * If community.foo does not relax its dependency on community.bar for some time, community.foo will be removed from the next Ansible major release.
>    * Unfortunately community.bar has to stay at 1.2.x until either community.bar is removed (in the next major release), or loosens its requirements so that newer community.foo 1.y.z releases can be included.
> 
> #. community.foonetwork depends on ansible.netcommon >= 2.0.0, <3.0.0.
> 
>    * ansible.netcommon 4.0.0 is released during this major Ansible release cycle.
>    * community.foonetwork either releases a new version before feature freeze of the next major Ansible release that allows depending on all ansible.netcommon 4.x.y releases, or it will be removed from the next major Ansible release.

Reviewer 2: ```suggestion
   * Now ``community.bar`` creates a new release ``1.3.0``. When ``community.foo`` does not create a new release with a relaxed dependency, we have to include ``community.bar 1.2.x`` in the next Ansible release despite ``1.3.0`` being available.
   * If ``community.foo`` does not relax its dependency on ``community.bar`` for some time, ``community.foo`` will be removed from the next Ansible major release.
   * Unfortunately ``community.bar`` has to stay at ``1.2.x`` until either community.bar is removed (in the next major release), or loosens its requirements so that newer ``community.foo 1.y.z`` releases can be included.

#. ``community.foonetwork`` depends on ``ansible.netcommon >= 2.0.0, <3.0.0``.

   * ``ansible.netcommon 4.0.0`` is released during this major Ansible release cycle.
   * ``community.foonetwork`` either releases a new version before feature freeze of the next major Ansible release that allows depending on all ``ansible.netcommon 4.x.y`` releases, or it will be removed from the next major Ansible release.
```

>    * ``community.foonetwork`` either releases a new version before feature freeze of the next major Ansible release that allows depending on all ``ansible.netcommon 4.x.y`` releases, or it will be removed from the next major Ansible release.

Reviewer 3: This should be determined automatically, and there should be some sort of a state-file holding this info between releases.

Author: This sounds pretty complicated for something that hopefully never happens, and if it does can be handled with issues and already preparing removal from the upcoming major version (as its `ansible.in` already exists in a new directory).
